Source/WebDriver/Session.cpp

```diff
@@ -3282,21 +3283,89 @@ void Session::emitEvent(const String& eventName, RefPtr<JSON::Object>&& body)
+    Subscription subscription = { subscriptionID, WTFMove(events), WTFMove(browsingContextIDs), WTFMove(userContextIDs) };
+    m_subscriptions.add(subscriptionID, WTFMove(subscription));
+    completionHandler(CommandResult::success(JSON::Value::create(subscriptionID)));
 }
 
-void Session::enableGlobalEvent(const String& eventName)
+void Session::unsubscribeByIDs(Vector<String>&& subscriptionIDs, Function<void(CommandResult&&)>&& completionHandler)
 {
-    m_globalEventSet.add(eventName);
+    for (const auto& id : subscriptionIDs) {
+        if (!m_subscriptions.contains(id)) {
+            completionHandler(CommandResult::fail(CommandResult::ErrorCode::InvalidArgument, "At least one subscription id is unknown"_s));
+            return;
+        }
+    }
+
+    for (const auto& id : subscriptionIDs)
+        m_subscriptions.remove(id);
+
+    completionHandler(CommandResult::success());
 }
 
-void Session::disableGlobalEvent(const String& eventName)
+void Session::unsubscribeByEventName(Vector<String>&& eventNames, Function<void(CommandResult&&)>&& completionHandler)
 {
-    m_globalEventSet.remove(eventName);
+    HashMap<String, Subscription> subscriptionsToKeep;
+    HashSet<String> matchedEvents;
+    // TODO Process/validate list of event names (e.g. expanding if given only the module name)
+    // https://bugs.webkit.org/show_bug.cgi?id=291371
+    for (const auto& eventName : eventNames) {
+        for (const auto& subscription : m_subscriptions) {
+            if (!subscription.value.events.contains(eventName)) {
+                subscriptionsToKeep.add(subscription.value.id, subscription.value);
+                continue;
+            }
+
+            // FIXME Add support to subscribe to specific browsing contexts
+            // https://bugs.webkit.org/show_bug.cgi?id=282981
+            // In this case, only process them if we were given the "contexts" parameter
+            if (!subscription.value.isGlobal()) {
+                subscriptionsToKeep.add(subscription.value.id, subscription.value);
+                continue;
+            }
+
+            auto currentSubscriptionEventNames = subscription.value.events;
+            currentSubscriptionEventNames.removeAll(eventName);
+            matchedEvents.add(eventName);
+            if (!currentSubscriptionEventNames.isEmpty()) {
+                auto clonedSubscription = subscription.value;
+                clonedSubscription.events = currentSubscriptionEventNames;
+                subscriptionsToKeep.add(clonedSubscription.id, WTFMove(clonedSubscription));
+            }
+        }
+    }
+
+    if (matchedEvents.size() != eventNames.size()) {
+        completionHandler(CommandResult::fail(CommandResult::ErrorCode::InvalidArgument));
+        return;
+    }
+
+    // Only modify the actual subscriptions after we validated all requested events.
+    m_subscriptions = WTFMove(subscriptionsToKeep);
+
+    // TODO Send the unsubscription request to the browser?
```

Comment:
Not sure if you meant to leave in this comment.

Reply:
Oops. Removed